go-controller/pkg/cni/cniserver_test.go

```diff
@@ -206,3 +208,80 @@ func TestCNIServer(t *testing.T) {
 		}
 	}
 }
+
+func TestCNIServerOverlappingRequests(t *testing.T) {
+	tmpDir, err := utiltesting.MkTmpdir("cniserver")
+	if err != nil {
+		t.Fatalf("failed to create temp directory: %v", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	socketPath := filepath.Join(tmpDir, serverSocketName)
+	fakeClient := fake.NewSimpleClientset()
+
+	firstAddStarted := make(chan bool)
+	const magicSandboxID string = "foobar"
+
+	s := NewCNIServer(tmpDir, fakeClient)
+	if err := s.Start(func(request *PodRequest, kclient kubernetes.Interface) ([]byte, error) {
+		if request.Command == CNIAdd {
+			if request.SandboxID == magicSandboxID {
+				close(firstAddStarted)
+				time.Sleep(10 * time.Second)
+			}
+			return []byte("{}"), nil
+		}
+		return nil, nil
+	}); err != nil {
+		t.Fatalf("error starting CNI server: %v", err)
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			Dial: func(proto, addr string) (net.Conn, error) {
+				return net.Dial("unix", socketPath)
+			},
+		},
+	}
+
+	makeRequest := func(cmd command, sandbox string) *Request {
+		return &Request{
+			Env: map[string]string{
+				"CNI_COMMAND":     string(cmd),
+				"CNI_CONTAINERID": sandbox,
+				"CNI_NETNS":       "/some/path",
+				"CNI_ARGS":        "K8S_POD_NAMESPACE=awesome-namespace;K8S_POD_NAME=awesome-name",
+			},
+			Config: []byte("{\"cniVersion\": \"0.1.0\",\"name\": \"ovnkube\",\"type\": \"ovnkube\"}"),
+		}
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(2)
+	var add1Code, add2Code, del1Code, del2Code int
+	go func() {
+		_, add1Code = clientDoCNI(t, client, makeRequest(CNIAdd, magicSandboxID))
+		_, del1Code = clientDoCNI(t, client, makeRequest(CNIDel, magicSandboxID))
+		wg.Done()
+	}()
+	go func() {
+		// Wait until the first add has been started by the handler
+		<-firstAddStarted
+		_, add2Code = clientDoCNI(t, client, makeRequest(CNIAdd, "blah"))
+		_, del2Code = clientDoCNI(t, client, makeRequest(CNIDel, "blah"))
+		wg.Done()
+	}()
+	wg.Wait()
```

Comment:
why running this in parallel?
is not this exercising the same scenario, but more deterministic?
```go
		_, add1Code = clientDoCNI(t, client, makeRequest(CNIAdd, magicSandboxID))
		<-firstAddStarted
		_, del1Code = clientDoCNI(t, client, makeRequest(CNIDel, magicSandboxID))
		_, add2Code = clientDoCNI(t, client, makeRequest(CNIAdd, "blah"))
		_, del2Code = clientDoCNI(t, client, makeRequest(CNIDel, "blah"))
```

Reply:
ic, you want to send the other commands meanwhile the magicsandboxId is in the cache

Reply:
maybe you can use 2 channels, one to signal that you've received the ADD and other to close when the other requests are done ... it will remove the sleep and make the test faster

```
done := make(chan struct{})
...
			if request.SandboxID == magicSandboxID {
				close(firstAddStarted)
				<-done
```